### PR TITLE
New octopus merge

### PR DESCRIPTION
**Pull request by litghost**
Merges:
 - https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/645
- https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/646
- https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/647
- https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/84

### CODE REVIEW

**Comment by acomodi** [comment]
@litghost I will try to get a detailed review ASAP. I see CI is red, I have tracked down which strong regression test is taking too much time and it appears to be `strong_titan/ucsb_152_tap_fir_stratixiv_arch_timing.blif`. I am still unsure though, on what is the specific issue.

**Comment by acomodi** [comment]
I also think it would be worth to use the same paradigm as the one that is used right now for this repo. Meaning, push on a branch(es) the new patch(es) and from there merge in `master+wip` once it is ready to go and open an issue tracking the branch. This way we can keep this fork fairly clean. What do you think?

**Comment by litghost** [comment]
> I also think it would be worth to use the same paradigm as the one that is used right now for this repo. Meaning, push on a branch(es) the new patch(es) and from there merge in `master+wip` once it is ready to go and open an issue tracking the branch. This way we can keep this fork fairly clean. What do you think?

It's unclear to me how what you are describing that is different from what I did?

**Comment by acomodi** [comment]
@litghost actually, creating a branch in symbiflow fork with the changes and create an issue to keep track of the patch, if it makes sense.

**Comment by litghost** [comment]
> @litghost actually, creating a branch in symbiflow fork with the changes and create an issue to keep track of the patch, if it makes sense.

That's what I did?

File: vpr/src/route/connection_box_lookahead_map.cpp [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+    float expected_delay = cost_entry.delay;
+    float expected_congestion = cost_entry.congestion;
+
+    float expected_cost = criticality_fac * expected_delay + (1.0 - criticality_fac) * expected_congestion;
+    return expected_cost;
+}
+
+/* runs Dijkstra's algorithm from specified node until all nodes have been
+ * visited. Each time a pin is visited, the delay/congestion information
+ * to that pin is stored is added to an entry in the routing_cost_map */
```
Comment (acomodi) [inline vpr/src/route/connection_box_lookahead_map.cpp]: `is added` is a leftover I suppose.

File: vpr/src/route/connection_box_lookahead_map.cpp [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+        const auto& device_ctx = g_vpr_ctx.device();
+        const std::vector<int>& channel_node_list = device_ctx.rr_node_indices[rr_type_][start_x_][start_y_][0];
+
+        for (; index_ < channel_node_list.size(); index_++) {
+            int node_ind = channel_node_list[index_];
+
+            if (device_ctx.rr_nodes[node_ind].capacity() == 0) {
+                continue;
+            }
+
```
Comment (acomodi) [inline vpr/src/route/connection_box_lookahead_map.cpp:226]: Here there should be a check to verify that the node_ind is OPEN. With the non-consecutive ptc that will be merged it will be needed.

File: vpr/src/route/connection_box_lookahead_map.cpp [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+ * have many cost entries at each |dx|,|dy| offset. There are many ways to
+ * "boil down" the many costs at each offset to a single entry for a given
+ * (wire type, chan_type) combination we can take the smallest cost, the
+ * average, median, etc. This define selects the method we use.
+ *
+ * See e_representative_entry_method */
+#define REPRESENTATIVE_ENTRY_METHOD SMALLEST
+
+#define REF_X 25
+#define REF_Y 23
```
Comment (acomodi) [inline vpr/src/route/connection_box_lookahead_map.cpp:24]: What are these numbers related to? Meaning, do they work well generally or are they tuned for specific cases?

File: vpr/src/route/connection_box_lookahead_map.cpp [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+        int count = 0;
+
+        int dx = 0;
+        int dy = 0;
+        //int start_x = vtr::nint(device_ctx.grid.width()/2);
+        //int start_y = vtr::nint(device_ctx.grid.height()/2);
+        int start_x = REF_X;
+        int start_y = REF_Y;
+        while ((count == 0 && dx < 7) || dy < 1) {
```
Comment (acomodi) [inline vpr/src/route/connection_box_lookahead_map.cpp]: How `7` and `1` are chosen? I suppose it would be better to use some defines.

File: vpr/src/route/router_lookahead_map_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+        /* skip this child if it has already been expanded from */
+        if (node_expanded[child_node_ind]) {
+            continue;
+        }
+
+        PQ_Entry child_entry(child_node_ind, switch_ind, parent_entry.delay,
+                             parent_entry.R_upstream, parent_entry.congestion_upstream, false);
+
+        //VTR_ASSERT(child_entry.cost >= 0);
```
Comment (acomodi) [inline vpr/src/route/router_lookahead_map_utils.cpp]: Why is this commented out?

File: utils/route_diag/src/main.cpp [hunk 6]
```diff
@@ -0,0 +1,269 @@
+// Tool to test routing from one node to another node.
+//
+// This tool is intended to allow extremely verbose router output on exactly
+// one path to diagnose router behavior, e.g. look ahead and A* characteristics.
```
Comment (acomodi) [inline utils/route_diag/src/main.cpp:4]: Probably worth adding a "How to use" example for this utility tool

**Review by acomodi — CHANGES_REQUESTED** [review]
LGTM apart from minor changes and clarifications. As far as I understood this PR introduces a new lookahead routing method which suits SymbiFlow needs and ish selectable by an option, right? This way the VTR tests remain unaffected by this change.

Does this work with `all_xc7` and `all_ice40` targets?

File: utils/route_diag/src/main.cpp (hunk 6 above)
Comment (litghost) [reply utils/route_diag/src/main.cpp:4]: Done

File: vpr/src/route/connection_box_lookahead_map.cpp (hunk 3 above)
Comment (litghost) [reply vpr/src/route/connection_box_lookahead_map.cpp:24]: They are tuned, but so are the look ahead REF_X/Y this code originated from, so I'm not super concerned with this at the outset.

File: vpr/src/route/connection_box_lookahead_map.cpp (hunk 2 above)
Comment (litghost) [reply vpr/src/route/connection_box_lookahead_map.cpp:226]: Right, done

File: vpr/src/route/connection_box_lookahead_map.cpp (hunk 4 above)
Comment (litghost) [reply vpr/src/route/connection_box_lookahead_map.cpp]: Done

File: vpr/src/route/connection_box_lookahead_map.cpp (hunk 1 above)
Comment (litghost) [reply vpr/src/route/connection_box_lookahead_map.cpp]: Fixed

**Review by acomodi — APPROVED** [review]
LGTM. 
CI is failing, but I think because of an infrastructural issue, some of the tests failed as they did not find the executables

**Comment by acomodi** [comment]
@litghost The CI failure is also happening upstream, so I believe we can merge this. I am about to push force on master+wip a cleaner version of the current octopus merge with all the open branches on top (`all_xc7` should finish in a couple of minutes). 
Can you please rebase as soon as I have pushed the latest changes?

**Comment by litghost** [comment]
> @litghost The CI failure is also happening upstream, so I believe we can merge this. I am about to push force on master+wip a cleaner version of the current octopus merge with all the open branches on top (`all_xc7` should finish in a couple of minutes).
> Can you please rebase as soon as I have pushed the latest changes?

Will do.  I assume all_ice40 passed already?

**Comment by acomodi** [comment]
> Will do. I assume all_ice40 passed already?

Yes indeed. I didn't check though with this PR, would you like me to locally run symbiflow tests with this PR merged locally on top of master+wip?

**Comment by litghost** [comment]
> > Will do. I assume all_ice40 passed already?
> 
> Yes indeed. I didn't check though with this PR, would you like me to locally run symbiflow tests with this PR merged locally on top of master+wip?

That would be appreciated.  Let me update this PR.  Have you seen an upstream issue on VTR about the Travis issue?  I'm wondering if it is a ccache failure?

**Comment by acomodi** [comment]
> That would be appreciated. Let me update this PR. Have you seen an upstream issue on VTR about the Travis issue? I'm wondering if it is a ccache failure?

No issue yet, I will open one. Anyways it started failing after presumably unrelated changes (https://github.com/verilog-to-routing/vtr-verilog-to-routing/commit/d016fffe50ab77f059dbea7c972cbf70a37f1fa8) so there is some issue with the infrastructure.

**Comment by litghost** [comment]
> presumably unrelated changes ([verilog-to-routing@d016fff](https://github.com/verilog-to-routing/vtr-verilog-to-routing/commit/d016fffe50ab77f059dbea7c972cbf70a37f1fa8)) so there is some issue with the infrastructure.

I saw that too, and made a mental note to take a look again later.

**Comment by litghost** [comment]
> LGTM.
> CI is failing, but I think because of an infrastructural issue, some of the tests failed as they did not find the executables

I've update the merge, can you take a look?  I'm going to kick off all_xc7 and all_ice40 and see if symbiflow passes.

**Comment by acomodi** [comment]
@litghost Some QoR of the regression tests have failed (I am reporting only the failures):
```
regression_tests/vtr_reg_strong/strong_dedicated_clock...[Fail] 
 timing/k6_frac_N10_frac_chain_mem32K_htree0_40nm.xml/verilog/mkPktMerge.v/common_--clock_modeling_dedicated_network num_routed_nets: golden = 950 result = 944
[Fail] 
 timing/k6_frac_N10_frac_chain_mem32K_htree0_routedCLK_40nm.xml/verilog/mkPktMerge.v/common_--clock_modeling_dedicated_network num_routed_nets: golden = 950 result = 944
[Fail] 
 timing/k6_frac_N10_frac_chain_mem32K_htree0short_40nm.xml/verilog/mkPktMerge.v/common_--clock_modeling_dedicated_network num_routed_nets: golden = 950 result = 944
regression_tests/vtr_reg_strong/strong_place_delay_model...[Fail] 
 stratixiv_arch.timing.xml/styr.blif/common_--place_delay_model_delta device_limiting_resources: golden = io LAB result = io
[Fail] 
 stratixiv_arch.timing.xml/styr.blif/common_--place_delay_model_delta_override device_limiting_resources: golden = io LAB result = io
regression_tests/vtr_reg_strong/strong_fpu_hard_block_arch...[Fail] 
 hard_fpu_arch_timing.xml/mm3.v/common num_post_packed_nets: golden = 289 result = 353
[Fail] 
 hard_fpu_arch_timing.xml/mm3.v/common device_limiting_resources: golden = io result = block_FPU
[Fail] 
 hard_fpu_arch_timing.xml/mm3.v/common pack_time: golden = 0.83 result = 8.90
[Fail] 
 hard_fpu_arch_timing.xml/mm3.v/common logic_block_area_used: golden = 68766.3 result = 103149

regression_tests/vtr_reg_strong/strong_power...[Fail] 
 k6_frac_N10_mem32K_40nm.xml/ch_intrinsics.v/common total_power: golden = 0.008868 result = 0.009777

regression_tests/vtr_reg_strong/strong_global_routing...[Fail] 
 nonuniform_chan_width/k6_N10_mem32K_40nm_nonuniform.xml/stereovision3.v/common min_chan_width: golden = 12 result = 16
```

Given the fact that the `connection_box_lookahead` is not used in the regression tests, one of the remaining causes could be the change in the placer delay matrix which could affect the QoR.
